src/Npgsql/NpgsqlCommand.cs

```diff
@@ -1143,16 +1143,24 @@ async ValueTask<NpgsqlDataReader> ExecuteReaderAsync(CommandBehavior behavior, b
                         if (connector.Settings.MaxAutoPrepare > 0)
                         {
                             var numPrepared = 0;
+
+                            var now = DateTime.UtcNow;
+
+                            // mark already prepared statements which are used in this batch, preventing them from being evicted in the second loop
+                            foreach (var statement in _statements)
```

Comment:
This adds another pass over statements, which is really something I'd like to avoid - this is pretty perf-sensitive code. Any idea why my attempted fix in #3145 isn't enough? Also, have you managed to reproduce the issue you're seeing in a test?

I'll try to take a deeper look at this again tomorrow.

Reply:
@roji you should be able to reproduce that on #3145 using the `DontAutoPrepareMoreThanMaxStatementsInBatchRandom` test from this PR